internal/resources/providers/gcplib/inventory/resource_manager.go

```diff
@@ -0,0 +1,142 @@
+	getProjectDisplayName      func(ctx context.Context, parent string) string
+	getOrganizationDisplayName func(ctx context.Context, parent string) string
+}
+
+func NewResourceManagerWrapper(ctx context.Context, log *clog.Logger, gcpConfig auth.GcpFactoryConfig) (*ResourceManagerWrapper, error) {
+	var gcpClientOpt []option.ClientOption
+	gcpClientOpt = append(append(gcpClientOpt, option.WithScopes(cloudresourcemanager.CloudPlatformReadOnlyScope)), gcpConfig.ClientOpts...)
+	crmService, err := cloudresourcemanager.NewService(ctx, gcpClientOpt...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ResourceManagerWrapper{
+		log:                  log,
+		config:               gcpConfig,
+		accountMetadataCache: sync.Map{},
+		getProjectDisplayName: func(ctx context.Context, parent string) string {
+			prj, err := crmService.Projects.Get(parent).Context(ctx).Do()
+			if err != nil {
+				log.Errorf("error fetching GCP Project: %s, error: %s", parent, err)
+				return ""
+			}
+			return prj.DisplayName
+		},
+		getOrganizationDisplayName: func(ctx context.Context, parent string) string {
+			org, err := crmService.Organizations.Get(parent).Context(ctx).Do()
+			if err != nil {
+				log.Errorf("error fetching GCP Org: %s, error: %s", parent, err)
+				return ""
+			}
+			return org.DisplayName
+		},
+	}, nil
+}
+
+func (c *ResourceManagerWrapper) GetCloudMetadata(ctx context.Context, asset *assetpb.Asset) *fetching.CloudAccountMetadata {
+	orgId := getOrganizationId(asset.Ancestors)
+	projectId := getProjectId(asset.Ancestors)
+	key := fmt.Sprintf("%s/%s", projectId, orgId)
+	cloudAccount, ok := c.accountMetadataCache.Load(key)
+	if ok {
+		cloudAccountMetadata, valid := cloudAccount.(*fetching.CloudAccountMetadata)
+		if valid {
+			return cloudAccountMetadata
+		}
+		c.log.Errorf("error casting cloud account metadata for key: %s", key)
+	}
+	cloudAccountMetadata := c.getMetadata(ctx, orgId, projectId)
+	c.accountMetadataCache.Store(key, cloudAccountMetadata)
+	return cloudAccountMetadata
+}
+
+func (c *ResourceManagerWrapper) getMetadata(ctx context.Context, orgId string, projectId string) *fetching.CloudAccountMetadata {
+	var orgName string
+	var projectName string
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		if isOrganization(c.config.Parent) {
+			orgName = c.getOrganizationDisplayName(ctx, fmt.Sprintf("organizations/%s", orgId))
+		}
+	}()
+	go func() {
+		defer wg.Done()
+		// some assets are not associated with a project
+		if projectId != "" {
+			projectName = c.getProjectDisplayName(ctx, fmt.Sprintf("projects/%s", projectId))
+		}
+	}()
+	wg.Wait()
+
+	return &fetching.CloudAccountMetadata{
+		AccountId:        projectId,
+		AccountName:      projectName,
+		OrganisationId:   orgId,
+		OrganizationName: orgName,
+	}
+}
+
+func (c *ResourceManagerWrapper) Clear() {
+	c.accountMetadataCache.Clear()
+}
+
+func getOrganizationId(ancestors []string) string {
+	last := ancestors[len(ancestors)-1]
```

Comment:
nit: perhaps a check if the ancestors has zero length (is it possible?)

Reply:
ancestors can't be empty
https://github.com/googleapis/google-cloud-go/blob/656d319252e5d973366a2c10f57f072b170dee1f/asset/apiv1p5beta1/assetpb/assets.pb.go#L95-L103